### ISSUE TITLE
Proposal: NgCache decorators|annotations & Zone cache implementation

### ISSUE THREAD

**Issue by apoterenko** [angular/angular#12472]
**I'm submitting a proposal.** 
Please consider the proposal, maybe you introduce something like that https://www.npmjs.com/package/angular2-cache

**Current behavior**
Is absent, therefore I have reinvented the wheel

**Expected behavior**
`export class Service {

    private id:string;             
    private expiration:string;      
	
    @ZoneCached()  	// The result will be cached within working zone
    public getExpirationDate():Date {
        return this.expiration
            ? new Date(this.expiration)
            : null;
    }
}

@Injectable()
export class NgZoneGlobalCache extends MemoryCache<any, any> {

    constructor(@Inject(NgZone) ngZone:NgZone) {
        super();

        ngZone.onStable.subscribe(() => {
            this.clear();  // Clear the memory cache when ngZone completed its work
        });

        GlobalCachesInstances.ZONE_INSTANCE = this;
    }
}`

**What is the motivation / use case for changing the behavior?**
I wanna get a solution out of the box.

* **Angular version:** 2.0.X

* **Browser:** all
 
* **Language:** TypeScript 1.x, 2.x

**Comment by robwormald** [comment]
I'm not entirely clear what the goal of this is - can you expand on it? Specifically, what Zone has to do with caching (or why you'd cache a Zone in the first place)

**Comment by apoterenko** [comment]
Consider the simple example:

`dateValuesArray = ['01/01/2010', '01/01/2010', '01/01/2010', ...., '01/01/2040']; // Array might contain duplicates, array.length = 20000;

Template:

<div *ngFor="dateValue in dateValuesArray">
   Date object: {{ parseValue(dateValue) }}
</div>

@Component(...) {

   @ZoneCached()
   getParsedDateValue(dateValue:string) {
       return new Date(dateValue);			// It called only once, not three times for '01/01/2010' value, the result of "new Date('01/01/2010')" should be cached in memory and 		
											// cleared when V8 "tick" life cycle has been finished => ngZone.onStable.subscribe(() => memoryCache.clear() });
   }
}
`

**Comment by vicb** [comment]
Could you please describe what you are trying to do in plain english. We do have enough spare time to spend 10mn trying to understand what people want on every single issue.

**Comment by apoterenko** [comment]
Ok, I have understood, the proposal needs a more detailed proposal.

**Comment by vicb** [comment]
The best would be to write a design spec (ie google docs). Thanks !

**Comment by robwormald** [comment]
If I'm understanding you correctly, you're trying to avoid calling parseValue repeatedly?

In general, you should avoid binding to methods, as they only way to check if a value has changed is to call said method. Using a pure pipe is another way to handle this.
